rb_link/src/server.rs

```diff
@@ -0,0 +1,284 @@
+use crate::config::*;
+use serde::{Deserialize, Serialize};
+use std::collections::{HashMap, VecDeque};
+use std::io::{Read, Write};
+use std::os::unix::net::{UnixListener, UnixStream};
+use std::sync::{Arc, Mutex};
+use std::thread::{self, JoinHandle};
+use std::time::Duration;
+use std::{fs, usize};
+
+#[derive(Serialize, Deserialize)]
+pub enum GetPutMessage {
+    Get(u32),
+    Put(B2RMessage),
+}
+
+#[derive(Serialize, Deserialize)]
+struct R2BMessage {
+    id: u32,
+    message: Vec<u8>,
+}
+
+/// Message from Bluesim:
+/// - id: ID of the probe that sent the message
+/// - cycles: Clock cycles when the message was sent
+/// - message: Sent message, where message.len() == ceil(put_t_width/8). put_t_width is the width of put_t defined in your BSV code.
+#[derive(Serialize, Deserialize)]
+pub struct B2RMessage {
+    pub id: u32,
+    pub cycles: u32,
+    pub message: Vec<u8>,
+}
+
+/// A server for interacting with Bluesim.
+/// Cache bidirectional data and send data upon receiving requests.
+pub struct B2RServer {
+    probe_infos: HashMap<ProbeType, Vec<u32>>,
+    b2r_cache: Arc<Mutex<HashMap<u32, VecDeque<B2RMessage>>>>,
+    r2b_cache: Arc<Mutex<HashMap<u32, VecDeque<R2BMessage>>>>,
+}
+
+/// Type of The Probe
+/// Fifo: won't get data from rust, sent 2 bytes every cycle, the fist byte is notFull second byte is notEmpty
+/// Fired: won't get data from rust, sent 1 bytes message when the rule fired
+#[derive(PartialEq, Eq, Hash)]
+pub enum ProbeType {
+    Fifo,
+    Fired,
+}
+
+/// The pipeline state
+/// cycle: the cycle of the state
+/// full_fifos: the ids of the full fifos
+/// empty_fifos: the ids of the empty fifos
+/// fire_fules: the fired rules
+pub struct PipeLineState {
+    pub cycle: u32,
+    pub full_fifos: Vec<u32>,
+    pub empty_fifos: Vec<u32>,
+    pub fire_rules: Vec<u32>,
+}
+
+impl Default for B2RServer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl B2RServer {
+    /// make a new server
+    pub fn new() -> B2RServer {
+        B2RServer {
+            probe_infos: HashMap::new(),
+            b2r_cache: Arc::new(Mutex::new(HashMap::new())),
+            r2b_cache: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+
+    // give a type to a probe so that it can be analysis by server.
+    pub fn give_type(&mut self, probe_type: ProbeType, id: u32) {
+        let ids = self.probe_infos.entry(probe_type).or_default();
+        if !ids.contains(&id) {
+            ids.push(id);
+        }
+    }
+
+    /// Start a thread to run the server.
+    /// Create a UnixListener at "/tmp/b2rr2b".
+    /// Return the JoinHandle of that thread.
+    /// This function needs to be called before running your Bluesim program.
+    pub fn serve(&mut self) -> JoinHandle<()> {
+        // let probe_infos = self.probe_infos.clone();
+        let b2r_cache = self.b2r_cache.clone();
+        let r2b_cache = self.r2b_cache.clone();
+        thread::spawn(move || {
```

Comment:
When will this thread terminated? what will happen if this thread terminated?

Reply:
fixed